### DISCUSSION TITLE
Issue with hanging nodes + small strain plasticity

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#24759] [Q&A General] [answered]
Hi all,

I am having issues when combining hanging nodes with a small strain plasticity material. This is not with adaptivity, the mesh starts with hanging nodes and does not change. 
For testing, I am running a problem in uniaxial tension. The mesh is made and refined using MOOSE. 
I have the following simulation parameters:

- Tensor mechanics module
- - Small strain + incremental formulation
- volumetric locking correction
- Isotropic elasticity
- Power law hardening

The refined area of the mesh is designed to yield first. 
I've attached the input file I used(I had to attach it as a .txt file).
[tension_moose.txt](https://github.com/idaholab/moose/files/11818976/tension_moose.txt)

While the plastic strain field looks ok, the stress field contains significant oscillations, as seen in the image attached. These oscillations are always concentrated near the transition zones for mesh refinement. 
![image](https://github.com/idaholab/moose/assets/75846521/cdd7f7e7-f979-4a29-a452-cbe0fd129cde)

I have tried this with several iterations on the problem setup (finite strain, no vol lock correction, total formulation) but I see the same stress oscillations each time. Also, I see no oscillations with a purely elastic material.
Any insight into this issue would be much appreciated.

**Comment by GiudGiud** [comment]
@lindsayad @dschwen I think this is tied to a recent conversation.

**Reply by TheGreatCid** [reply]
Is there an incompatibility with how data interpolation works on hanging nodes and plasticity?

**Reply by lindsayad** [reply]
Do you see oscillations without adaptivity?

**Reply by TheGreatCid** [reply]
There is no adaptivity, the mesh intentionally starts with hanging nodes

**Reply by TheGreatCid** [reply]
But using unstructured quad mesh refinement does not lead to oscillations

**Reply by GiudGiud** [reply]
Adaptivity is not used btw. It s the refineSidesetGenerator.

Does it work with a single level of refinement?

**Reply by TheGreatCid** [reply]
![image](https://github.com/idaholab/moose/assets/75846521/29090545-9a1a-4b70-9dd8-5c8184326d4b)
The oscillations are still present with 1 level of refinement

**Reply by GiudGiud** [reply]
ok can you try with adaptivity instead of sideset refinement now?

Use a constant marker boundary-restricted to the boundary of interest

I apologize in advance if that does not work, I m trying to see if we can get your case going as our user support resources are low right now

**Reply by TheGreatCid** [reply]
![image](https://github.com/idaholab/moose/assets/75846521/85d96a9d-f29f-46ec-9749-94884421567b)
Unfortunately, that gives similar oscillations

**Reply by lindsayad** [reply]
It seems like we should start making a list of physics that don't play well with hanging nodes :-/

**Reply by dschwen** [reply]
What? How? How can this be physics specific? Can you show us the displacement field? Are you sure your solution is tightly converged? The residuals at the hanging nodes seem to converge from opposite signs, so a badly converged solution would exhibit exactly this checkerboarding as far as I can see.

**Reply by lindsayad** [reply]
Along @dschwen's line of questioning, @TheGreatCid can you paste a representative timestep solve history?

**Reply by TheGreatCid** [reply]
You make a salient point. My abs and ref convergence criterion were 1e-8. I increased the ref to be 1e-15 and abs to be 1e-10 to make sure it converged on abs. It has some trouble reaching convergence, but the steps that do converge do not have the checkerboarding pattern. So what you said is most certainly correct. 
![image (11)](https://github.com/idaholab/moose/assets/75846521/24d03a26-28fd-41eb-8a4c-5551149eca17)

**Reply by TheGreatCid** [reply]
Here is a representative step with my original tolerances
![image](https://github.com/idaholab/moose/assets/75846521/77a487ab-19f8-4974-9e08-e5e67ea781c1)

Here is a typical step with the tightened tolerance
![image](https://github.com/idaholab/moose/assets/75846521/d2248a9c-e493-4a22-b6c7-0b375f89f52d)

So this is very much indicative of the fact that I was not reaching a fully converged solution

**Reply by lindsayad** [reply]
It's very nice to see the improved solution!

On a different note, the nonlinear convergence is very poor. I see that you are using NEWTON. Are you confident the Jacobian is good? How is convergence if you use PJFNK and add
```
[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]
```

**Reply by TheGreatCid** [reply]
Hmm, if I make the switch PJFNK with that preconditioner I get the following output:
![image](https://github.com/idaholab/moose/assets/75846521/19c21e08-ab7f-4149-961d-07bd779e18f6)
So it seems to cause issues, though I do not know what DIVERGED_BREAKDOWN is referring to

**Reply by lindsayad** [reply]
Can you try with `-pc_type lu`?

**Reply by hugary1995** [reply]
@lindsayad this is the old tensor mechanics material with inexact Jacobian.

**Reply by TheGreatCid** [reply]
> Can you try with `-pc_type lu`?

It invokes a similar resposne

**Reply by lindsayad** [reply]
> @lindsayad this is the old tensor mechanics material with inexact Jacobian.

Thanks @hugary1995. Yea I just wanted to eliminate the possible effect of the constraints on boomeramg. So given what @hugary1995 said and the demonstrably bad linear convergence with PJFNK and nonlinear convergence with NEWTON, we know the Jacobian is poor. You should probably use the automatic differentiation version of the mechanics kernels or the Lagrangian kernels that @reverendbedford created

**Reply by maxnezdyur** [reply, marked as answer]
@TheGreatCid 
The picture below is after 100 seconds. Two main things, using AD and `preset = false` for your BC. I am also using LU. `preset=false` is something that I found really helps. I think the oscillations are gone if you look at the colorbar bounds.
```
E = 207e3

[GlobalParams]
  displacements = 'disp_x disp_y'
  volumetric_locking_correction = true
[]
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 1
    ymax = 10
    nx = 4
    ny = 40
  []
  [block]
    type = SubdomainBoundingBoxGenerator
    block_id = 5
    block_name = 'ref'
    input = gmg
    bottom_left = '-2 4 0'
    top_right = '2 6 0'
  []
  [refine]
    type = RefineBlockGenerator
    input = block
    block = 5
    refinement = 2
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = SMALL
    incremental = true
    generate_output = 'stress_yy'
    use_automatic_differentiation = true
    decomposition_method = EigenSolution
  []
[]

[Materials]
  [elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = ${E}
    poissons_ratio = 0.3
    block = 0
  []
  [power_law_hardening]
    type = ADIsotropicPowerLawHardeningStressUpdate
    strength_coefficient = 0.1 #K
    strain_hardening_exponent = 0.5 #n
    block = 0
  []
  [radial_return_stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'power_law_hardening'
    # tangent_operator = elastic
    block = 0
  []
  [elasticity_tensormid]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = ${E}
    poissons_ratio = 0.3
    block = 5
  []
  [power_law_hardeningmid]
    type = ADIsotropicPowerLawHardeningStressUpdate
    strength_coefficient = 0.1 #K
    strain_hardening_exponent = 0.5 #n
    block = 5
    output_properties = effective_plastic_strain
    outputs = exodus
  []
  [radial_return_stressmid]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'power_law_hardeningmid'
    # tangent_operator = elastic
    block = 5
  []
[]

[BCs]
  [left_x]
    type = DirichletBC
    value = 0
    boundary = left
    variable = disp_x
  []
  [bot_y]
    type = DirichletBC
    value = 0
    boundary = bottom
    variable = disp_y
  []
  [top]
    type = FunctionDirichletBC
    function = '0.01*t'
    boundary = top
    variable = disp_y
    preset = false
  []
[]
[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]
[Executioner]
  type = Transient
  end_time = 100
  solve_type = NEWTON
  line_search = none
  dtmin = 1e-20
  nl_max_its = 50
  nl_rel_tol = 1e-11
  nl_abs_tol = 1e-9
  [TimeStepper]
    type = ConstantDT
    dt = 1
  []
  petsc_options = '-ksp_snes_ew'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu superlu_dist'
  automatic_scaling = false
[]
[Outputs]
  print_linear_residuals = false
  exodus = true
  interval = 1
[]
```
![image](https://github.com/idaholab/moose/assets/22859100/dd44b8b5-5c4b-4db3-b17d-20fcb3e95439)

**Reply by TheGreatCid** [reply]
> You should probably use the automatic differentiation version of the mechanics kernels 
@lindsayad That makes sense, and as @maxnezdyur showed, appears to be the solution here
